src/Data/Foreign.purs

```diff
@@ -44,16 +44,20 @@ foreign import data Foreign :: *
 
 -- | A type for runtime type errors
 data ForeignError
-  = TypeMismatch String String
+  = TypeMismatch (Array String) String
   | ErrorAtIndex Int ForeignError
   | ErrorAtProperty String ForeignError
   | JSONError String
 
 instance showForeignError :: Show ForeignError where
-  show (TypeMismatch exp act) = "Type mismatch: expected " <> exp <> ", found " <> act
   show (ErrorAtIndex i e) = "Error at array index " <> show i <> ": " <> show e
   show (ErrorAtProperty prop e) = "Error at property " <> show prop <> ": " <> show e
   show (JSONError s) = "JSON error: " <> s
+  show (TypeMismatch exps act) = "Type mismatch: expected " <> to_s exps <> ", found " <> act
+    where
+      to_s [] = "???"
```

Comment:
I agree, but just to say, `purescript-nonempty` is in contrib currently and this is core.